File: scraper/buscar-olx.mjs
import * as cheerio from "cheerio";

// User-Agent precisa parecer um navegador de verdade — um UA minimalista tipo
// "Mozilla/5.0" sozinho leva 403 do OLX (testado e confirmado em produção).
const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

export async function buscarOlx(config) {
  const url = `https://www.olx.com.br/celulares/celulares-e-smartphones?q=${encodeURIComponent(config.termos_busca)}`;
  const res = await fetch(url, {
    headers: {
      "User-Agent": USER_AGENT,
      Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
      "Accept-Language": "pt-BR,pt;q=0.9",
    },
  });

  if (!res.ok) {
    throw new Error(`OLX respondeu ${res.status} pra "${config.termos_busca}"`);
  }

  const html = await res.text();
  const $ = cheerio.load(html);

  const precoMin = config.preco_min ?? 0;
  const precoMax = config.preco_max ?? Infinity;
  const resultados = [];

  $(".olx-adcard").each((_, el) => {
    const linkEl = $(el).find('[data-testid="adcard-link"]').first();
    const titulo = linkEl.find(".olx-adcard__title").text().trim() || linkEl.attr("title")?.trim() || "";
    const link = linkEl.attr("href");

    const precoTexto = $(el).find(".olx-adcard__price").first().text();
    const preco = Number(precoTexto.replace(/\D/g, ""));

    const localizacao = $(el)
      .find(".olx-adcard__location")
      .first()
      .text()
      .replace(/\s+/g, " ")
      .trim();

    const imagem = $(el).find(".olx-adcard__media img").first().attr("src") ?? null;

    if (!link) return;
    const externalId = link.split("-").pop()?.split("?")[0];
    if (!externalId || !preco) return;
    if (preco < precoMin || preco > precoMax) return;

    resultados.push({
      id: externalId,
      titulo,
      preco,
      link,
      imagem,
      descricao: "",
      localizacao: localizacao || config.localizacao || null,
    });
  });

  return resultados;
}

File: scraper/buscar_olx.py
"""
Busca no OLX via Scrapling (StealthyFetcher — navegador real disfarçado).

O OLX roda atrás do Cloudflare, que bloqueia requisição HTTP simples (mesmo com
User-Agent de navegador) vinda de IP de datacenter. O StealthyFetcher do Scrapling
resolve o desafio JS do Cloudflare de verdade — testado e confirmado funcionando
tanto local quanto no runner do GitHub Actions.
"""

from urllib.parse import quote
from scrapling.fetchers import StealthyFetcher
from util_localizacao import bate_cidade


def _texto(selector_list):
    if not selector_list:
        return ""
    return "".join(selector_list[0].css("::text").getall()).strip()


def buscar_olx(config: dict) -> list[dict]:
    termos = config["termos_busca"]
    url = f"https://www.olx.com.br/celulares/celulares-e-smartphones?q={quote(termos)}"

    page = StealthyFetcher.fetch(url, headless=True, solve_cloudflare=True, network_idle=True)
    if page.status != 200:
        raise RuntimeError(f'OLX respondeu {page.status} pra "{termos}"')

    preco_min = config.get("preco_min") or 0
    preco_max = config.get("preco_max") or float("inf")
    cidades = config.get("cidades") or []
    resultados = []

    for card in page.css(".olx-adcard"):
        link_el = card.css('[data-testid="adcard-link"]')
        if not link_el:
            continue
        link = link_el[0].attrib.get("href")
        if not link:
            continue

        titulo = _texto(link_el[0].css(".olx-adcard__title")) or (link_el[0].attrib.get("title") or "").strip()

        preco_texto = _texto(card.css(".olx-adcard__price"))
        digitos = "".join(c for c in preco_texto if c.isdigit())
        preco = int(digitos) if digitos else None

        localizacao = _texto(card.css(".olx-adcard__location"))

        img_el = card.css(".olx-adcard__media img")
        imagem = img_el[0].attrib.get("src") if img_el else None

        external_id = link.rstrip("/").split("-")[-1].split("?")[0]
        if not external_id or not preco:
            continue
        if preco < preco_min or preco > preco_max:
            continue
        if not bate_cidade(localizacao, cidades):
            continue

        resultados.append(
            {
                "id": external_id,
                "titulo": titulo,
                "preco": preco,
                "link": link,
                "imagem": imagem,
                "descricao": "",
                "localizacao": localizacao or None,
            }
        )

    return resultados
